frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java

```diff
@@ -61,42 +59,34 @@ public void removePlaceVisit(String tripId, String placeId) throws PlaceVisitNot
 
   @Override
   public Optional<PlaceVisitModel> getPlaceVisit(String tripId, String placeId) {
-    return Optional.ofNullable(storage.get(tripId)).map(placesMap -> placesMap.get(placeId));
+    return Optional.ofNullable(placesByPlaceIdByTripId.get(tripId))
+        .map(placesMap -> placesMap.get(placeId));
   }
 
   @Override
   public boolean updateUserMarkOrAddPlaceVisit(
       PlaceVisitModel placeVisit, PlaceVisitModel.UserMark newStatus) {
+    AtomicBoolean alreadyInStorage = new AtomicBoolean(false);
 
-    boolean[] alreadyInStorage = new boolean[1];
-    alreadyInStorage[0] = false;
+    Map<String, PlaceVisitModel> placesMap = placesByPlaceIdByTripId.computeIfAbsent(
+        placeVisit.tripId(), (tripKey) -> new ConcurrentHashMap<>());
 
-    storage.compute(placeVisit.tripId(), (tripKey, placesMap) -> {
-        if (placesMap == null) {
-          Map<String, PlaceVisitModel> newPlaceMap = new ConcurrentHashMap<>();
-          newPlaceMap.put(placeVisit.placeId(), placeVisit);
-          return newPlaceMap;
-        }
-        
-        if (placesMap.get(placeVisit.placeId()) != null) {
-          PlaceVisitModel updatedPlace = placesMap.get(placeVisit.placeId()).toBuilder()
-            .setUserMark(newStatus)
-            .build();
-          placesMap.put(updatedPlace.placeId(), updatedPlace);
-          alreadyInStorage[0] = true;
-        } else {
-          placesMap.put(placeVisit.placeId(), placeVisit);
-        }
-        
-        return placesMap;
+    placesMap.compute(placeVisit.placeId(), (placeKey, place) -> {
+      if (place != null) {
+        PlaceVisitModel updatedPlace = place.toBuilder().setUserMark(newStatus).build();
+        alreadyInStorage.set(true);
+        return updatedPlace;
+      } else {
+        return placeVisit;
```

Comment:
Does the given place visit have the current user mark in this case? If so, feel free to resolve this. If not, I assume you’ll need to set the mark here.